File: Reporting Tool/hg_api.py
import os
import time
import hg_commands as commands
from commit import Commit

class HgApi:
	'''
		Wraps the mercurial commands in a convenient api 
	'''
	
	def __init__(self):
		pass

	def _execute(self, call_string):
		handle = os.popen(call_string)
		return handle.read()

	def list_all_branches(self):
		return self._execute(commands.list_all_branches)[:-1]

	# Executes an 'hg pull' in cwd
	def pull(self):
		self._execute(commands.pull)

	def checkout_branch(self, branch_name):
		self._execute(commands.update + " " + branch_name)

	def checkout_commit(self, commit_hash):
		self._execute(commands.checkout_commit + ' ' + commit_hash)

	def last_commit_date(self, branch_name):
		return self._execute(commands.last_commit_date + " " + branch_name).split()[0]

	def revert(self):
		self._execute(commands.revert)

	"""
		Returns a list of Commit objects (see commit.py) with attributes (rev_id, commit_hash, commit_date)
	"""
	def get_commit_list(self):
		commit_list = self._execute(commands.commit_list).split("\n")[:-1]
		commits = []
		for commit in commit_list:
			commit_info = commit.split("_")
			new_commit = Commit(commit_info[0], commit_info[1], commit_info[-1].split("-")[0])
			commits.append(new_commit)
		return commits

File: reporting_DSL/src/main/scala/reportingDSL/HgApi.scala
package reportingDSL

object HgApi extends Execute {
  type Branch = String

  def listAllBranches(repository: String): Seq[String] = {
    executeInDirectory(HgCommands.listAllBranches, repository)
  }

  def pull(repository: String): Seq[String] = {
    executeInDirectory(HgCommands.pull, repository)
  }

  def checkoutBranch(repository: String, branch: Branch): Seq[String] = {
    executeInDirectory(HgCommands.update + " " + branch, repository)
  }

  def checkoutCommit(repository: String, commit: String): Seq[String] = {
    executeInDirectory(HgCommands.update + " " + commit, repository)
  }

  def getLastCommitDate(repository: String, branch: Branch): String = {
    executeInDirectory(HgCommands.lastCommitDate + " " + branch, repository)(0).split(" ")(0).substring(1)
  }

  def revert(repository: String): Unit = {
    executeInDirectory(HgCommands.revert, repository)
  }

  def getCommitList(repository: String): Seq[Commit] = {
    executeInDirectory(HgCommands.commitList, repository).dropRight(1).map(commit => {
      val commitInfo = commit.replace(""""""", "").split("_")
      Commit(revisionId = commitInfo(0).trim(), hash = commitInfo(1), date = commitInfo(2).split(" ")(0))
    })
  }
}